crates/pathfinder/src/p2p_network/sync_handlers/v1.rs

```diff
@@ -0,0 +1,579 @@
+) -> anyhow::Result<()> {
+    let responses = spawn_blocking_get(request, storage, events).await?;
+    send(tx, responses).await
+}
+
+pub(crate) fn headers(
+    tx: Transaction<'_>,
+    request: BlockHeadersRequest,
+) -> anyhow::Result<Vec<BlockHeadersResponse>> {
+    let BlockHeadersRequest {
+        iteration:
+            Iteration {
+                start_block,
+                direction,
+                limit,
+                step,
+            },
+    } = request;
+
+    let mut next_block_number = BlockNumber::new(start_block);
+    let mut limit = limit.min(MAX_BLOCKS_COUNT);
+
+    let mut responses = Vec::new();
+
+    while let Some(block_number) = next_block_number {
+        if limit == 0 {
+            break;
+        }
+
+        // 1. Get the block header
+        let Some(header) = tx.block_header(block_number.into())? else {
+            // No such block
+            break;
+        };
+
+        responses.push(BlockHeadersResponse {
+            block_number: block_number.get(),
+            header_message: BlockHeaderMessage::Header(Box::new(header.to_proto())),
+        });
+
+        // 2. Get the signatures for this block
+        // TODO we don't have signatures yet
+
+        limit -= 1;
+        next_block_number = get_next_block_number(block_number, step, direction);
+    }
+
+    Ok(responses)
+}
+
+fn bodies(
+    tx: Transaction<'_>,
+    request: BlockBodiesRequest,
+) -> anyhow::Result<Vec<BlockBodiesResponse>> {
+    let BlockBodiesRequest {
+        iteration:
+            Iteration {
+                start_block,
+                direction,
+                limit,
+                step,
+            },
+    } = request;
+
+    let mut next_block_number = BlockNumber::new(start_block);
+    let mut limit = limit.min(MAX_BLOCKS_COUNT);
+
+    let mut responses = Vec::new();
+
+    while let Some(block_number) = next_block_number {
+        if limit == 0 {
+            break;
+        }
+
+        // 1. Get the state diff for the block
+        let Some(state_diff) = tx.state_update(block_number.into())? else {
+            // No such block
+            break;
+        };
+
+        let new_classes = state_diff
+            .declared_cairo_classes
+            .iter()
+            .copied()
+            .chain(
+                state_diff
+                    .declared_sierra_classes
+                    .keys()
+                    .map(|x| ClassHash(x.0)),
+            )
+            .collect::<Vec<_>>();
+
+        responses.push(BlockBodiesResponse {
+            block_number: block_number.get(),
+            body_message: BlockBodyMessage::Diff(state_diff.to_proto()),
+        });
+
+        // 2. Get the newly declared classes in this block
+        let get_compressed_definition =
+            |block_number: BlockNumber, class_hash| -> anyhow::Result<Vec<u8>> {
+                let definition = tx
+                    .compressed_class_definition_at(block_number.into(), class_hash)?
```

Comment:
Is the compression type used in the spec the same as in our DB?

Reply:
It's not specified right now so I just left it as is